default/AI/ResearchAI.py

```diff
@@ -453,28 +492,45 @@ def generate_research_orders():
     rng = random.Random()
     rng.seed(fo.getEmpire().name + fo.getGalaxySetupData().seed)
 
-    research_reqs = calculate_research_requirements(empire)
+    calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
     if total_rp <= 0: # No RP available - no research.
         return
 
+    raw_priorities = {}
     priorities = {}
+    on_path_to = {}
     for tech_name in fo.techs():
-        priority = get_priority(rng, tech_name)
+        this_tech = fo.getTech(tech_name)
+        if not this_tech:
+            continue
+        raw_priorities[tech_name] = get_priority(rng, tech_name)
+    for tech_name, priority in raw_priorities.iteritems():
         if not tech_is_complete(tech_name) and priority >= 0:
-            turn_needed = float(research_reqs[tech_name][1]) / total_rp
-            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
+            turn_needed = max(research_reqs[tech_name][2], math.ceil(float(research_reqs[tech_name][1]) / total_rp))
+            adjusted_priority = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
About the many staged tech line-- yes, I had started to do a first inheritance of priorities that only had a portion of the time scaling applied to it, and then at the end the prereqs would have the additional time scaling factor applied based on their own individual times to completion.  Something about that implementation had apparently gone off track, though, and the end priorities were much smaller than I was expecting.  I didn't really have time to sort it out right now so I set it aside.  But I think some approach along those lines will probably be the best way to work in the inherited priorities.  I also want to experiment a bit with using an exponential factor like you had previously suggested.   I'm not so crazy about the aspect of the current formula that treats any finish time within the next 12 turns as equivalent.

As far as cost vs time, I think they're both important.  The cost is factored in here, but I would agree that by rolling it into the time factor it will often be buried.  So, yes, a formula with a greater role for cost would seem to have good prospects.